stan/math/rev/mat/functor/algebra_system.hpp

```diff
@@ -0,0 +1,156 @@
+     * @tparam T0 type for unknowns
+     * @tparam T1 type for auxiliary parameters
+     * @tparam x_is_iv true if x is the independent variable
+     */
+    template <typename F, typename T0, typename T1, bool x_is_iv>
+    struct system_functor {
+      F f_;  /** algebraic system functor */
+      Eigen::Matrix<T0, Eigen::Dynamic, 1> x_;  /** unknowns */
+      Eigen::Matrix<T1, Eigen::Dynamic, 1> y_;  /** auxiliary parameters */
+      std::vector<double> dat_;  /** real data */
+      std::vector<int> dat_int_;  /** integer data */
+      std::ostream* msgs_;  /** stream message */
+
+      system_functor() { }
+
+      system_functor(const F& f,
+                     const Eigen::Matrix<T0, Eigen::Dynamic, 1>& x,
+                     const Eigen::Matrix<T1, Eigen::Dynamic, 1>& y,
+                     const std::vector<double>& dat,
+                     const std::vector<int>& dat_int,
+                     std::ostream* msgs)
+        : f_(), x_(x), y_(y), dat_(dat), dat_int_(dat_int), msgs_(msgs) { }
+
+      /**
+       * An operator that takes in an independent variable. The
+       * independent variable is either passed as the unknown x,
+       * or the auxiliary parameter y. The x_is_iv template parameter
+       * allows us to determine whether the jacobian is computed
+       * with respect to x or y.
+       * @tparam T the scalar type of the independent variable
+       */
+      template <typename T>
+      inline
+      Eigen::Matrix<T, Eigen::Dynamic, 1>
+      operator()(const Eigen::Matrix<T, Eigen::Dynamic, 1>& iv) const {
+        if (x_is_iv)
+          return f_(iv, y_, dat_, dat_int_, msgs_);
+        else
+          return f_(x_, iv, dat_, dat_int_, msgs_);
+      }
+    };
+
+    /**
+     * A structure which gets passed to Eigen's dogleg
+     * algebraic solver.
+     * @tparam T scalar type of independent variable.
+     * @tparam NX number of rows
+     * @tparam NY number of columns
+     */
+    template <typename T, int NX = Eigen::Dynamic, int NY = Eigen::Dynamic>
+    struct nlo_functor {
+      const int m_inputs, m_values;
+
+      nlo_functor() : m_inputs(NX),
+                      m_values(NY) { }
+
+      nlo_functor(int inputs, int values) : m_inputs(inputs),
+                                            m_values(values) { }
+
+      int inputs() const { return m_inputs; }
+      int values() const { return m_values; }
+    };
+
+    /**
+     * A functor with the required operators to call Eigen's 
+     * algebraic solver.
+     * @tparam FS wrapper around the algebraic system functor. Has the
+     * signature required for jacobian (i.e takes only one argument).
+     * @tparam F algebraic system functor
+     * @tparam T0 scalar type for unknowns
+     * @tparam T1 scalar type for auxiliary parameters
+     */
+    template <typename FS, typename F, typename T0, typename T1>
+    struct hybrj_functor_solver : nlo_functor<double> {
+      FS fs_;  /** Wrapper around algebraic system */
```

Comment:
Doxygen doc like this is attached to the following declaration.  So you want to move these up one.